File: app/src/assessment/scoring.ts
import type { Question, Answer, State } from './types';

const VALID: Answer[] = ['yes', 'no', 'na'];

export function deriveState(
  questions: Pick<Question, 'qid' | 'informs_state'>[],
  answers: Record<string, Answer | null | undefined>,
): State {
  // Empty is a programming error (archetype UCs always have questions; bespoke is
  // scored manually and never reaches here) — throw rather than silently fall
  // through to MET, which would mark an unassessed use case as fully covered.
  if (questions.length === 0) throw new Error('deriveState requires at least one question');
  const vals = questions.map(q => [q.informs_state, answers[q.qid] ?? null] as const);
  if (vals.length && vals.every(([, v]) => v === 'na')) return 'NA';
  if (vals.some(([, v]) => v == null || !VALID.includes(v as Answer))) return 'PENDING';
  if (vals.some(([inf, v]) => inf === 'GAP_PARTIAL' && v === 'no')) return 'GAP';
  if (vals.some(([inf, v]) => inf === 'PARTIAL_MET' && v === 'no')) return 'PARTIAL';
  return 'MET';
}

File: methodology/scoring.py
"""Reference posture-state derivation for the WS-1 archetype rubric.

derive_state applies the laddering rule for an archetype use case (A1-A8).
A0 (bespoke) use cases are scored manually and do NOT call this function.
Pure: no I/O, no globals. Mirrored by questionnaire/scoring.js."""

VALID_ANSWERS = {"yes", "no", "na"}


def derive_state(questions, answers):
    """questions: [{"qid": str, "informs_state": "GAP_PARTIAL"|"PARTIAL_MET"}].
    answers: {qid: "yes"|"no"|"na"|None}. Returns GAP|PARTIAL|MET|PENDING|NA.

    Empty `questions` is a programming error (an archetype UC always has questions;
    A0/bespoke is scored manually and never calls this) — raise rather than silently
    fall through to MET, which would mark an unassessed use case as fully covered."""
    if not questions:
        raise ValueError("derive_state requires at least one question "
                         "(empty input would silently return MET)")
    vals = [(q["informs_state"], answers.get(q["qid"])) for q in questions]
    if vals and all(v == "na" for _, v in vals):
        return "NA"
    if any(v not in VALID_ANSWERS for _, v in vals):   # None or missing
        return "PENDING"
    if any(inf == "GAP_PARTIAL" and v == "no" for inf, v in vals):
        return "GAP"
    if any(inf == "PARTIAL_MET" and v == "no" for inf, v in vals):
        return "PARTIAL"
    return "MET"
